### PR TITLE
feat(unifyassumes): verifier.assert->sea.assert.if using unified-assumes

### PR DESCRIPTION
**Pull request by priyasiddharth**
each verifier.assert is converted to sea.assert.if using unified-assumes seeon so far
Also added verifier.assert.not
Also added a flag to turn sassert into verifier.assert + existing sassert

### CODE REVIEW

File: include/seahorn/Expr/Smt/Z3.hh [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
  */
 
+#include "seahorn/Expr/ExprCore.hh"
```
Comment (agurfinkel) [inline include/seahorn/Expr/Smt/Z3.hh]: why is this needed?

File: include/seahorn/Expr/Smt/Z3.hh [hunk 2]
```diff
@@ -587,6 +588,14 @@ public:
       *(out++) = z3.toExpr(r[i]);
   }
 
+  /// Dumps assertions currently in the solver to stderr
+  void dump() const {
+    z3::ast_vector r(ctx, Z3_solver_get_assertions(ctx, solver));
+    ctx.check_error();
+    for (unsigned i = 0; i < r.size(); ++i)
+      std::cerr << *(z3.toExpr(r[i])) << std::endl;
```
Comment (agurfinkel) [inline include/seahorn/Expr/Smt/Z3.hh]: `llvm::errs()` instead of `std::errs()`

If this is the only reason for including `ExprCore.hh` then move it to a cc file

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 3]
```diff
@@ -81,15 +95,35 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
   auto *assumeFn = SBI.mkSeaBuiltinFn(seahorn::SeaBuiltinsOp::ASSUME, M);
 
   SmallVector<CallInst *, 16> assumes;
+  // an orphan assert is one which does not have a preceding assume
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc]: I am very confused by this code. Marking "to discuss"

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 4]
```diff
@@ -105,9 +139,21 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
   assumeFlag = B.CreateAlloca(B.getInt1Ty(), 0U, nullptr, "assume.flag");
   B.CreateStore(B.getTrue(), assumeFlag);
 
+  // -- process all orphan asserts first
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc]: Don't understand the concept of `orphan` assert, so marking anything related to it as to discuss

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 5]
```diff
@@ -135,6 +181,34 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
   return false;
 }
 
+void UnifyAssumesPass::processAssertInst(CallInst &CI, AllocaInst &flag) {
+  BasicBlock *bb = CI.getParent();
+  Module *M = bb->getParent()->getParent();
+  assert(bb);
+  assert(M);
+
+  auto &SBI = getAnalysis<seahorn::SeaBuiltinsInfoWrapperPass>().getSBI();
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc]: make `SBI` a field and instantiate it during `runOnFunction`

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 6]
```diff
@@ -135,6 +181,34 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
...
+  llvm::CallSite CS(&CI);
+  IRBuilder<> B(bb);
+  B.SetInsertPoint(&CI);
+  auto ante = B.CreateLoad(&flag);
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc:179]: proper terminology should be `antecedent` and `consequent` not `antecedent` and `condition`

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 7]
```diff
@@ -135,6 +181,34 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
...
+    cond = B.CreateNot(cond);
+  }
+  // transform verifier.assert(cond) into sea.assert.if(flag, cond)
+  Function *assertIfFn =
+      SBI.mkSeaBuiltinFn(seahorn::SeaBuiltinsOp::ASSERT_IF, *M);
+  CallInst *NewCI = CallInst::Create(assertIfFn, {ante, cond});
+  NewCI->setCallingConv(assertIfFn->getCallingConv());
+  if (!CI.use_empty())
+    CI.replaceAllUsesWith(NewCI);
+  ReplaceInstWithInst(&CI, NewCI);
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc:193]: check whether this moved debug location from old to new instruction

File: lib/seahorn/BmcPass.cc [hunk 8]
```diff
@@ -362,6 +362,43 @@ class BmcPass : public llvm::ModulePass {
       }
     }
 
+    // -- compute dependence of verifier.assume.not()
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc]: comment and code disagree. Comment talks about `assume` but code about `assert.if`. Yet `assumeFn` variable indicates that this is an `assume` and not an `assert`

File: lib/seahorn/BmcPass.cc [hunk 9]
```diff
@@ -362,6 +362,43 @@ class BmcPass : public llvm::ModulePass {
...
+      for (auto *u : assumeFn->users()) {
+        if (auto *CI = dyn_cast<CallInst>(u)) {
+          CallSite CS(CI);
+          if (CS.getCaller() != &F)
+            continue;
+          dfCoi.analyze(*CI);
+        }
+      }
+    }
+    // -- compute dependence of verifier.assume.not()
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc]: same comment/code disagreement

**Review by agurfinkel — CHANGES_REQUESTED** [review]
something is fishy in unify assumes. Otherwise all is good

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 6 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:179]: fixed

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 7 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:193]: Yes. Code comment:
/// Replace the instruction specified by From with the instruction specified by
/// To. Copies DebugLoc from BI to I, if I doesn't already have a DebugLoc.

File: lib/seahorn/BmcPass.cc (hunk 9 above)
Comment (priyasiddharth) [reply lib/seahorn/BmcPass.cc]: fixed.

File: lib/seahorn/BmcPass.cc (hunk 8 above)
Comment (priyasiddharth) [reply lib/seahorn/BmcPass.cc]: fixed

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 5 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: passing it explicitly to callee since best practice with class field would suggest setting it at object creation time.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 4 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: removed. uneeded.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: removed

File: include/seahorn/Expr/Smt/Z3.hh (hunk 2 above)
Comment (priyasiddharth) [reply include/seahorn/Expr/Smt/Z3.hh]: removing this code for now from this PR.

File: include/seahorn/Expr/Smt/Z3.hh (hunk 1 above)
Comment (priyasiddharth) [reply include/seahorn/Expr/Smt/Z3.hh]: removing this code for now from this PR.

**Comment by priyasiddharth** [comment]
Responded to review comments... ok for next round of review/approval.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 5 above)
Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: No, this is an LLVM idiom. `runOnFunction` is essentially a creation method, analyses are extracted at that time and stored in fields.

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 10]
```diff
@@ -57,6 +59,18 @@ bool isAssumeCall(const CallInst &ci, seahorn::SeaBuiltinsInfo &SBI) {
     return true;
   }
 }
+
+bool isAssertCall(const CallInst &ci, seahorn::SeaBuiltinsInfo &SBI) {
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc]: as discussed, move  `SBI` to `m_SBI` field and set it in `runOnFunction`

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 11]
```diff
@@ -81,15 +95,23 @@ bool UnifyAssumesPass::runOnFunction(Function &F) {
   auto *assumeFn = SBI.mkSeaBuiltinFn(seahorn::SeaBuiltinsOp::ASSUME, M);
 
   SmallVector<CallInst *, 16> assumes;
+  SmallVector<CallInst *, 16> asserts;
+  // TODO: is this efficient
+  DenseMap<CallInst *, SmallVector<CallInst *, 16>> assumeToAssert;
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc]: is `assumeToAssert` still used?

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 7 above)
Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc:193]: I bet `ReplaceInstWithInst` also takes care of updating all the users, so lines `185--186` are not necessary.

File: lib/seahorn/BmcPass.cc [hunk 12]
```diff
@@ -362,6 +362,43 @@ class BmcPass : public llvm::ModulePass {
...
+          CallSite CS(CI);
+          if (CS.getCaller() != &F)
+            continue;
+          dfCoi.analyze(*CI);
+        }
+      }
+    }
+    // -- compute dependence of verifier.assert.not()
+    assumeFn = m->getFunction("verifier.assert.not");
+    if (assumeFn) {
```
Comment (agurfinkel) [inline lib/seahorn/BmcPass.cc]: this did not change. still called `assumeFn`. I think this should be factored out into a method that takes a pointer to `llvm::Function`, and if the pointer is not null computes its dependence. Then the method is called 3 times in here.

**Review by agurfinkel — COMMENTED** [review]
Some comments and single comments, some are as this review. Sorry for the mess.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 10 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: I don't know if you saw my previous comment - https://github.com/priyasiddharth/seahorn/pull/2#discussion_r475896092

File: lib/seahorn/BmcPass.cc (hunk 12 above)
Comment (priyasiddharth) [reply lib/seahorn/BmcPass.cc]: Fixed naming for now.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 11 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: Removed.

File: lib/seahorn/BmcPass.cc (hunk 12 above)
Comment (priyasiddharth) [reply lib/seahorn/BmcPass.cc]: We anyway wanted to get away from using strings to using defined enums... can this be in the next PR? Added in project task.

**Comment by priyasiddharth** [comment]
PTAL. Added .clang-format in test dir ... see commit msg for details

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 10 above)
Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: https://github.com/priyasiddharth/seahorn/pull/2#discussion_r475901061

Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: see my reply to your reply

File: test/.clang-format [hunk 13]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (agurfinkel) [inline test/.clang-format:1]: you really should not sneak it as part of a completely different commit, but ok

Comment (priyasiddharth) [reply test/.clang-format:1]: Ack

File: lib/seahorn/BmcPass.cc (hunk 12 above)
Comment (priyasiddharth) [reply lib/seahorn/BmcPass.cc]: refactored

File: test/.clang-format (hunk 13 above)
Comment (priyasiddharth) [reply test/.clang-format:1]: made into a separate commit.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 10 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: Fixed

**Comment by priyasiddharth** [comment]
Fixed all previous comments AFAICT. PTAL.

**Comment by priyasiddharth** [comment]
I will go ahead and merge with my local branch. You can still review in the `assert_if` branch
